keps/sig-node/3953-node-resource-hot-plug/README.md

```diff
@@ -0,0 +1,832 @@
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+- Change in OOMScoreAdjust value:
+    - The formula to calculate OOMScoreAdjust is `1000 - (1000*containerMemReq)/memoryCapacity`
+    - With change in memoryCapacity post up-scale, The existing OOMScoreAdjust may not be inline with the
+      actual OOMScoreAdjust for existing pods.
+        - This can be mitigated by recalculating the OOMScoreAdjust value for the existing pods. However, there can be an associated overhead for
+          recalculating the scores.
+- Change in Swap limit:
+    - The formula to calculate the swap limit is `<containerMemoryRequest>/<nodeTotalMemory>)*<totalPodsSwapAvailable>`
+    - With change in nodeTotalMemory and totalPodsSwapAvailable post up-scale, The existing swap limit may not be inline with the
+      actual swap limit for existing pods.
+        - This can be mitigated by recalculating the swap limit for the existing pods. However, there can be an associated overhead for
+          recalculating the scores.
+
+- Post up-scale any failure in resync of Resource managers may be lead to incorrect or rejected allocation, which can lead to underperformed or rejected workload.
+  - To mitigate the risks adequate tests should be added to avoid the scenarios where failure to resync resource managers can occur.
+
+- Lack of coordination about change in resource availability across kubelet/runtime/plugins.
+  - The plugins/runtime should be updated to react to change in resource information on the node.
+
+- Kubelet missing hotplug event or too many hotplug events
+  - Hotplug events are captured via periodic polling by the kubelet, this ensures that the capacity is updated in the poll cycle and can technically not miss the event/fail to handle a flood of events.
+
+- Handling downsize events
+  - Though there is no support through this KEP to handle an event of node-downsize, it's the onus of the cluster administrator to resize responsibly to avoid disruption as it lies out of the kubernetes realm.
+  - However, enabling this feature will ensure that the correct resource information is pushed across the cluster.
+
+## Design Details
+
+Below diagram is shows the interaction between kubelet, node and cAdvisor.
+
+
+```mermaid
+sequenceDiagram
+    participant node
+    participant kubelet
+    participant cAdvisor-cache
+    participant machine-info
+    kubelet->>cAdvisor-cache: Poll
+    cAdvisor-cache->>machine-info: fetch
+    machine-info->>cAdvisor-cache: update
+    cAdvisor-cache->>kubelet: update
+    kubelet->>node: node status update
+    kubelet->>node: re-initialize resource managers
+    kubelet->>node: Recalculate and update OOMScoreAdj <br> and Swap limit of pods
+```
+
+The interaction sequence is as follows
+1. Kubelet will be polling in interval to fetch the machine resource information from cAdvisor's cache, Which is currently updated every 5 minutes.
```

Comment:
Note that with an external signal about the current VM size coming from NRI Kubelet could maybe react within ~1s as opposed to the current 5 minutes of cadvisor cache refresh (see my comment in Motivation).

Reply:
True, but cAdvisor is currently using default poll interval of 5 min, which can be customized through `update_machine_info_interval` for aggressive polling. On the similar lines, if needed we can customize poling interval in kubelet through a flag.

Reply:
Sure, we will explore and update as per your suggestion https://github.com/kubernetes/enhancements/pull/3955#discussion_r1955119053